.github/workflows/testing-and-deployment.yml

```diff
@@ -53,7 +53,14 @@ jobs:
         run: pytest -v --ignore=tests/plotting
 
       - name: Test Plotting
-        run: pytest -v tests/plotting --fail_extra_image_cache
+        run: pytest -v tests/plotting --fail_extra_image_cache --generated_image_dir _debug_images
+
+      - name: Upload Images for Failed Tests
+        if: ${{ failure() && hashFiles('_debug_images/*.png') != '' }}
+        uses: actions/upload-artifact@v4
+        with:
+          name: debug-images
```

Comment:
I think we should have different naming for each job here.  Otherwise, if multiple jobs try to upload, it will throw an error.

Reply:
yes good idea. I'll see if there's a way to get the name of the job and insert it

Reply:
The artifact names now follow the convention:
`debug_images-PR-number_system-name_python-version_vtk-version`

Reply:
What happens if this is executed not from a PR event, e.g. via cron?  Since every action is already unique to a specific event, I don't think it is necessary to have the PR part, and might cause more problems.

Reply:
We can remove the 'PR' text and just keep the event number? I imagined working across multiple PRs and suddenly having a bunch of downloads with the same name. Or maybe needing to look for a generated image that you recall downloading a few commits ago and don't want to wade through the CI checks history. The PR number in this case would help find a particular debug image locally I think.

But no other artifacts in other workflows include the PR number, and it's probably best to actually review the full CI logs directly to make sure you are reviewing the correct set of images... So we can revert https://github.com/pyvista/pyvista/pull/6204/commits/9fcc69666c91b5d7c9da2f526f45e1549d38efaa

Reply:
I'm +1 on everything but the PR#.  I don't think the other events have a `github.event.number` and would be an error.

Reply:
We need to add the numpy version too for the 2.0 pre-release testing. This is added in https://github.com/pyvista/pyvista/pull/6204/commits/23f82b6d33c772e14e8f731004188c68c13d56a2 for the linux tests only. 

So the linux test artifact names now follow this convention:
`debug-images_system-name_python-version_vtk-version_numpy-version`

And windows/macos/linux-conda follow this convention:
`debug-images_system-name_python-version_vtk-version`

Reply:
Looked into this more.  I think we should use {{github.job}}.  This ensures these stay aligned with the job name and we don't have to maintain it 

https://docs.github.com/en/actions/learn-github-actions/contexts#github-context

Reply:
> Looked into this more. I think we should use {{github.job}}.

By itself unfortunately this doesn't give the name we want. By itself this evaluates to `windows` or `linux`, which is not a unique name:
Conflict: an artifact with this name already exists on the workflow run

I'm looking for a way to easily grab the job name as seen in the sidebar of the CI checks, e.g. the full name `Linux Unit Testing (3.12, latest, pre)`